service/internal/graph/graph.go

```diff
@@ -98,15 +105,70 @@ func (g *Graph) createNodes(set Settings) {
 		}
 	}
 
-	for connID, exprPipelineIDs := range connectorsAsExporter {
-		for _, eID := range exprPipelineIDs {
+	for connID := range connectors {
+		factory := set.ConnectorBuilder.Factory(connID.Type())
+		if factory == nil {
+			return fmt.Errorf("connector factory not available for: %q", connID.Type())
+		}
+		connFactory := factory.(connector.Factory)
+
+		expTypes := make(map[component.DataType]bool)
+		for _, pipelineID := range connectorsAsExporter[connID] {
+			// The presence of each key indicates how the connector is used as an exporter.
+			// The value is initially set to false. Later we will set the value to true *if* we
+			// confirm that there is a supported corresponding use as a receiver.
+			expTypes[pipelineID.Type()] = false
+		}
+		recTypes := make(map[component.DataType]bool)
+		for _, pipelineID := range connectorsAsReceiver[connID] {
+			// The presence of each key indicates how the connector is used as a receiver.
+			// The value is initially set to false. Later we will set the value to true *if* we
+			// confirm that there is a supported corresponding use as an exporter.
+			recTypes[pipelineID.Type()] = false
+		}
+
+		for expType := range expTypes {
+			for recType := range recTypes {
+				if connFactory.Stability(expType, recType) != component.StabilityLevelUndefined {
+					expTypes[expType] = true
+					recTypes[recType] = true
+				}
+			}
+		}
+
+		for expType, supportedUse := range expTypes {
+			if supportedUse {
+				continue
+			}
+			// Retain previous error message for special case where connector is used once, but incorrectly.
+			// TODO remove this case if we agree the slightly more generic error message is acceptable.
+			if len(expTypes) == 1 && len(recTypes) == 1 {
+				for recType := range recTypes {
+					return fmt.Errorf("connector %q cannot connect from %s to %s: telemetry type is not supported", connID, expType, recType)
+				}
+			}
```

Comment:
I would prefer to remove this and migrate to the slightly more generic error message seen below, but would appreciate other opinions on whether this is informative enough.

Reply:
I think it would be fine to remove this special case, we can always add it back if people find the generic error message confusing.

Reply:
Good catch. Artifacts from #8003. I've removed all changes to this file.

Reply:
I've removed the special case and updated expected error messages in tests.